src/version.rs

```diff
@@ -0,0 +1,263 @@
+                    .resolve(&*self.default_catalog, &*self.default_schema);
+                let full_object_name = ObjectName(vec![
+                    Ident::new(resolved_ref.catalog),
+                    Ident::new(resolved_ref.schema),
+                    Ident::new(resolved_ref.table),
+                ]);
+                if !self.rewrite_ready {
+                    // We haven't yet fetched/triaged the table versions ids; for now just collect
+                    // all table raw versions specified.
+
+                    self.table_versions
+                        .insert((full_object_name, value.to_string()), None);
+                } else {
+                    // Do the actual name rewrite
+                    name.0.last_mut().unwrap().value =
+                        self.table_with_version(&full_object_name, value);
+                    // Void the function table arg struct to leave a clean printable statement
+                    *args = None;
+                }
+            }
+        }
+        visit_table_table_factor(self, name, alias, args, with_hints)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::data_types::TableVersionId;
+    use datafusion::sql::parser::Statement;
+    use sqlparser::ast::Statement as SQLStatement;
+    use std::ops::Deref;
+    use test_case::test_case;
+
+    use crate::datafusion::parser::DFParser;
+    use crate::datafusion::visit::VisitorMut;
+    use crate::version::TableVersionProcessor;
+
+    #[test_case(
+        "SELECT * FROM test_table";
+        "Basic select with bare table name")
+    ]
+    #[test_case(
+        "SELECT * FROM some_schema.test_table";
+        "Basic select with schema and table name")
+    ]
+    #[test_case(
+        "SELECT * FROM some_db.some_schema.test_table";
+        "Basic select with fully qualified table name")
+    ]
+    fn test_table_version_rewrite(query: &str) {
+        let version_timestamp = "2022-01-01 20:01:01Z";
+        let stmts =
+            DFParser::parse_sql(format!("{}('{}')", query, version_timestamp).as_str())
+                .unwrap();
+
+        let mut q = if let Statement::Statement(stmt) = &stmts[0] {
+            if let SQLStatement::Query(query) = stmt.deref() {
+                query.clone()
+            } else {
+                panic!("Expected Query not matched!");
+            }
+        } else {
+            panic!("Expected Statement not matched!");
+        };
+
+        let mut rewriter = TableVersionProcessor::new(
+            "test_catalog".to_string(),
+            "test_schema".to_string(),
+        );
+        // Do a first pass
+        rewriter.visit_query(&mut q);
+
+        // Set a mock table version id and do another pass for actually renaming the table
+        let id = 42 as TableVersionId;
+        for table_version_id in rewriter.table_versions.values_mut() {
+            *table_version_id = Some(id);
+        }
+        rewriter.rewrite_ready = true;
```

Comment:
Just to confirm my understanding, during normal operation this is done by `triage_version_ids()`?

Reply:
That is correct.